artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -249,7 +253,7 @@
 
    private volatile boolean directDeliver = true;
 
-   private volatile boolean supportsDirectDeliver = true;
+   private volatile boolean supportsDirectDeliver = false;
```

Comment:
this should be default true.

Reply:
@clebertsuconic has proposed this change, to avoid having a small window in which a protocol that isn't suportting directDelivery could attempt to do one due to a race while adding its consumers

Reply:
fair enough.

Reply:
Yep, I see that I this PR the only relevant commit is the one that force MQTT and Stomp to avoid directDelivery, the rest are just improvements to avoid bad behaviours to happen..

Reply:
So i don't get why MQTT can have direct delivery here still, it used to, all you've done is found a bug it seems down in Core queue, which if thats the case, it will be a bug for all protocols, not just MQTT or STOMP.

Reply:
In theory it could happen on core too but it depends on "unlucky" timings: for core anyway I need to check the ordering of locking ServerConsumerImpl.this and QueueImpl.this to be sure of it.

Reply:
I understand this is settled, but let me explain with my words, to make sure you understand.. and I'm not wrong.

The reason we disable direct deliver is certain protocols won't have the locking support to deal with direct delivery, and they may dead lock.

So, with MQTT what I believe it could happen is.. supports= true, now you add a consumer, it becomes false.

If an user, out of an anti-pattern creates and deletes consumers rapidly while sending messages this could lead to races on delivering.

This is just for that kind of anti-pattern.

Reply:
so the deadlock isn't in MQTT from my understanding, its in QueueImpl. Its just MQTT was the protocol that the person using was on. The bug is there not in MQTT.

Reply:
My understanding was using direct delivery on a protocol that can't handle it.

Reply:
I'm not sure though.. this issue was reported Jan last year.. I thought I had worked on it after then. I didn't know it was still an issue until Franz worked on it today.


I didn't close the original JIRA before because the user never reported what version he was working on... 

So, I guess it's important to find a test replicating the issue, as we may be fixing something that was already fixed.

Reply:
@clebertsuconic if thats the case, then all the changes franz has done in QueueImpl are NOT needed. But looking through it does seem the issue IS in QueueImpl not MQTT.

Reply:
> So, I guess it's important to find a test replicating the issue, as we may be fixing something that was already fixed.

Agreed that is very important!!

Reply:
> I'm not sure though.. this issue was reported Jan last year.. I thought I had worked on it after then. I didn't know it was still an issue until Franz worked on it today.
> 
> I didn't close the original JIRA before because the user never reported what version he was working on...
> 
> So, I guess it's important to find a test replicating the issue, as we may be fixing something that was already fixed.

Hi,
I'm the guy who uploaded the thread dump to the jira issue last friday. Just want to make sure it's clear that we experienced this dead lock with version 2.6.4. Maybe I should've opened a new issue because it might have nothing to do with mqtt and it might be a different issue than the one the original reporter encountered a year ago with an older version.

Reply:
@michaelandrepearce @clebertsuconic 

According to @FrederikP it is an actual issue and with MQTT + debugging I've been able to reproduce it with ease: I just need to do the same using a test (and probably ByteMan or ByteBuddy to make it happen).
Re the why it can be fixed is not ease: according to what Clebert said I believe him that there are protocols that do not support direct delivery and MQTT is reporting (on the code) to support it. 
The simplest fix here is contained in one of the commit of this PR: just making it to report that it doesn't support direct delivery.

Thanks to the way MQTT works I have found that direct deliveries and async deliveries could happen concurrently (and TBH it doesn't make sense, from many point of views) AND that if the locks of ServerConsumerImpl and QueueImpl are not acquired in the correct order it could lead to a deadlock. 
Given that async and direct deliveries shouldn't happen concurrently (are mutually exclusive, but this wasn't never enforced on code level!), fixing it will make irrelevent the order of locks.

Now I'm checking if Core is exposed to the same issue.

Reply:
I think we need to have a reproducer test case here before making any assumptions or move forwards with reviewing.